core/src/main/java/org/polypheny/db/catalog/entity/logical/LogicalKey.java

```diff
@@ -0,0 +1,144 @@
+public abstract class LogicalKey implements PolyObject, Comparable<LogicalKey> {
+
+    @Serial
+    private static final long serialVersionUID = -5803762884192662540L;
+
+    @Serialize
+    public long id;
+
+    @Serialize
+    public long entityId;
+
+    @Serialize
+    public long namespaceId;
+
+    @Serialize
+    public ImmutableList<Long> columnIds;
+
+    @Serialize
+    public EnforcementTime enforcementTime;
+
+
+    public LogicalKey( long id, long entityId, long namespaceId, List<Long> columnIds, EnforcementTime enforcementTime ) {
+        this.id = id;
+        this.entityId = entityId;
+        this.namespaceId = namespaceId;
+        this.columnIds = ImmutableList.copyOf( columnIds );
+        this.enforcementTime = enforcementTime;
+    }
+
+
+    public String getSchemaName() {
+        return Catalog.snapshot().getNamespace( namespaceId ).orElseThrow().name;
+    }
+
+
+    public String getTableName() {
+        return Catalog.snapshot().rel().getTable( entityId ).orElseThrow().name;
+    }
+
+
+    public List<String> getColumnNames() {
+        Snapshot snapshot = Catalog.snapshot();
+        List<String> columnNames = new LinkedList<>();
+        for ( long columnId : columnIds ) {
+            columnNames.add( snapshot.rel().getColumn( columnId ).orElseThrow().name );
+        }
+        return columnNames;
+    }
+
+
+    @Override
+    public PolyValue[] getParameterArray() {
+        return new PolyValue[]{ PolyLong.of( id ), PolyLong.of( entityId ), PolyString.of( getTableName() ), PolyLong.of( namespaceId ), PolyString.of( getSchemaName() ), null, null };
+    }
+
+
+    @Override
+    public int compareTo( @NotNull LogicalKey o ) {
+        return (int) (this.id - o.id);
+    }
+
+
+    public enum EnforcementTime {
+        ON_COMMIT,
+        ON_QUERY
+    }
+
+
+    @Override
+    public boolean equals( Object o ) {
+        if ( this == o ) {
+            return true;
+        }
+        if ( o == null || getClass() != o.getClass() ) {
+            return false;
+        }
+
+        LogicalKey that = (LogicalKey) o;
+
+        if ( id != that.id ) {
+            return false;
+        }
+        if ( entityId != that.entityId ) {
+            return false;
+        }
+        if ( namespaceId != that.namespaceId ) {
+            return false;
+        }
+        if ( !columnIds.equals( that.columnIds ) ) {
+            return false;
+        }
+        return enforcementTime == that.enforcementTime;
+    }
+
+
+    @Override
+    public int hashCode() {
+        int result = (int) (id ^ (id >>> 32));
+        result = 31 * result + (int) (entityId ^ (entityId >>> 32));
+        result = 31 * result + (int) (namespaceId ^ (namespaceId >>> 32));
+        result = 31 * result + columnIds.hashCode();
+        result = 31 * result + (enforcementTime != null ? enforcementTime.hashCode() : 0);
+        return result;
+    }
```

Comment:
Please document why you have opt for implementing the hashCode and equals manually instead of using lombok.

Reply:
This an artifact of testing, as lombok does no like if one of hashcode() or equals function is manually implemented, when you use the annoations. Removed both and went back to annotation.